### PR TITLE
Add graceful termination for custom builders

### LINKED ISSUE #2656: Graceful termination for custom builder
### Expected behavior
When cancelling skaffold build with a custom builder, it would be helpful if the build command process had time to clean up. For example, first sending SIGTERM, and then waiting before sending SIGKILL.

### Actual behavior
Cancelling CommandContext sends SIGKILL immediately, preventing the custom builder from cleaning up. An example of something that might need to be cleaned up is a pod from a cluster builder.

### Information

- Skaffold version: v0.34.1
- Contents of skaffold.yaml:

```yaml
apiVersion: skaffold/v1beta13
kind: Config
build:
  artifacts:
  - image: ...
    custom:
      buildCommand: ./build.sh
```

### Steps to reproduce the behavior

1. `skaffold build` with any `build.sh`
2. Ctrl-C before `build.sh` completes

### PR DESCRIPTION
**Pull request by cedrickring**
Fixes #2656

I added a 2 seconds grace period for the process to complete after receiving SIGINT.

### CODE REVIEW

**Comment by tejal29** [comment]
@cedrickring, i feel the better way to do this is by defining a different cleanUpCommand.

```
apiVersion: skaffold/v1beta13
kind: Config
build:
  artifacts:
  - image: ...
    custom:
      buildCommand: ./build.sh
      cleanUpCommand: ./clean-up.sh
```
And then, skaffold to execute the `cleanUpCommand` and not wait for a constant time to cleanup.
What do you think?

**Comment by cedrickring** [comment]
Hey @tejal29, thanks for the feedback.

The (up to) 30 seconds wait is not exactly for cleanup imo, but for the command to exit gracefully. Right now the process gets killed immediately after pressing Ctrl+C and doesn't have time to exit normally (temp files aren't cleaned up etc). Usually a process shouldn't take the whole 30 seconds to terminate.
 
I think a cleanup script could be an additional option for this, but not a replacement.

File: pkg/skaffold/build/custom/custom_test.go [hunk 1]
```diff
@@ -125,8 +130,98 @@ func TestRetrieveCmd(t *testing.T) {
...
+		{
+			description: "terminate gracefully and kill process",
+			artifact: &latest.Artifact{
+				ArtifactType: latest.ArtifactType{
+					CustomArtifact: &latest.CustomArtifact{
+						BuildCommand: fmt.Sprintf("%s --test.run=TestHelperGracefulBuildKill", os.Args[0]),
+					},
+				},
+			},
+			testEnv:   []string{"GO_STARTED_FROM_TEST=1"},
```
Comment (tejal29) [inline pkg/skaffold/build/custom/custom_test.go]: this should unset in a `defer`.

File: pkg/skaffold/build/custom/custom_test.go [hunk 2]
```diff
@@ -125,8 +130,98 @@ func TestRetrieveCmd(t *testing.T) {
...
+			description: "terminate gracefully and exit 0",
+			artifact: &latest.Artifact{
+				ArtifactType: latest.ArtifactType{
+					CustomArtifact: &latest.CustomArtifact{
+						BuildCommand: fmt.Sprintf("%s --test.run=TestHelperGracefulBuildCancel", os.Args[0]),
+					},
+				},
+			},
+			testEnv:   []string{"GO_STARTED_FROM_TEST=1"},
+			shouldErr: false,
```
Comment (tejal29) [inline pkg/skaffold/build/custom/custom_test.go]: not needed as default is `false`

**Comment by tejal29** [comment]
@cedrickring, thanks for clarifying. I see, you are implementing what https://github.com/golang/go/issues/22757 proposes. 

I think, if we implement this solution, we should reduce the timeout for 2 seconds like 
https://github.com/mvdan/sh/pull/170/files#diff-b071a7aa58a1e42f9c5278440e585c77R49

File: pkg/skaffold/build/custom/custom_test.go [hunk 3]
```diff
@@ -125,8 +130,98 @@ func TestRetrieveCmd(t *testing.T) {
...
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+
+	go func() {
+		<-signals
+		fmt.Println("Got interrupt")
+	}()
+
+	fmt.Println("Sleeping 25 seconds")
```
Comment (tejal29) [inline pkg/skaffold/build/custom/custom_test.go]: i am little bit concerned with these tests adding sleeps. Can you just add a bash command like `sleep 5` in the build script?

Also, run it multiple times (100) to make sure its reliable.

Comment (cedrickring) [reply pkg/skaffold/build/custom/custom_test.go]: Imo 100 repeats are a bit much for a unit test. Also inlining a bash script required some extra changes in code.

Comment (cedrickring) [reply pkg/skaffold/build/custom/custom_test.go]: Also the tests require the interrupt signal to be trapped. Otherwise they will always error.

Comment (tejal29) [reply pkg/skaffold/build/custom/custom_test.go]: sorry, i meant running it 100 times via `go run -n -test.run <>`. I will do it myself before reviewing.

File: pkg/skaffold/build/custom/custom_test.go [hunk 4]
```diff
@@ -125,8 +139,57 @@ func TestRetrieveCmd(t *testing.T) {
...
+			},
+			shouldErr: true,
+		},
+	}
+
+	for _, test := range tests {
+		testutil.Run(t, test.description, func(t *testutil.T) {
+			builder := NewArtifactBuilder(false, test.testEnv)
+
+			for i := 0; i < 10; i++ {
```
Comment (tejal29) [inline pkg/skaffold/build/custom/custom_test.go]: please remove 10 iterations. Sorry I meant running it locally to make sure its not flaky.

File: pkg/skaffold/build/custom/custom.go [hunk 5]
```diff
@@ -52,19 +56,61 @@ func NewArtifactBuilder(pushImages bool, additionalEnv []string) *ArtifactBuilde
...
+	return err
 }
 
-func (b *ArtifactBuilder) retrieveCmd(ctx context.Context, out io.Writer, a *latest.Artifact, tag string) (*exec.Cmd, error) {
+func (b *ArtifactBuilder) retrieveCmd(out io.Writer, a *latest.Artifact, tag string) (*exec.Cmd, error) {
 	artifact := a.CustomArtifact
 	split := strings.Split(artifact.BuildCommand, " ")
 
-	cmd := exec.CommandContext(ctx, split[0], split[1:]...)
+	cmd := exec.Command(split[0], split[1:]...)
```
Comment (tejal29) [inline pkg/skaffold/build/custom/custom.go]: We don't want to support running arbitrary `bash scripts`.
Can we instead refactor `retriveCmd` and add a func `func buildCmd(string) *exec.Cmd`?

``` 
// for testing
var (
buildCmdFunc = buildCmd
)

func buildCmd(s string) *exec.Cmd {
    split := strings.Split(s " ")
     return exec.CommandContext(split[0], split[1:]...)
}

```

And in tests

```
func  mockBuildCmd(s string) (*exec.Cmd, error) {
split := strings.Split(s, " ")
 args := strings.Join(split[2:], " ")
 return  exec.Command(split[0], split[1], args[1:len(args)-1])
}




func TestGracefulBuildCancel(t *testing.T) {
...
 for _, test := range(tests) {
   testutil.Override(&retrieveCmdFunc, &mockBuildCmd)
  ....
}
```

File: pkg/skaffold/build/custom/custom.go [hunk 6]
```diff
@@ -52,19 +56,61 @@ func NewArtifactBuilder(pushImages bool, additionalEnv []string) *ArtifactBuilde
...
 	}
 
-	return cmd.Run()
+	if err := cmd.Start(); err != nil {
+		return errors.Wrap(err, "starting cmd")
+	}
+
+	done := make(chan struct{})
+	go func() {
+		select {
```
Comment (tejal29) [inline pkg/skaffold/build/custom/custom.go]: This code block is really long. Can we refactor this in to `handleGracefulTermination`

File: pkg/skaffold/build/custom/custom.go (hunk 5 above)
Comment (cedrickring) [reply pkg/skaffold/build/custom/custom.go]: This is obsolete after refactoring

File: pkg/skaffold/build/custom/custom.go [hunk 7]
```diff
@@ -52,19 +56,57 @@ func NewArtifactBuilder(pushImages bool, additionalEnv []string) *ArtifactBuilde
...
-	return cmd.Run()
+	if err := cmd.Start(); err != nil {
+		return errors.Wrap(err, "starting cmd")
+	}
+
+	return b.handleGracefulTermination(ctx, cmd)
+}
+
+func (b *ArtifactBuilder) handleGracefulTermination(ctx context.Context, cmd *exec.Cmd) error {
+	done := make(chan struct{})
```
Comment (tejal29) [inline pkg/skaffold/build/custom/custom.go]: This looks really great. Just one question, i am assuming you have the done channel to not send kill signal if the process is already done? Lets put a defer close as soon as you create the channel

We can also shorten this code block further

```
func (b *ArtifactBuilder) handleGracefulTermination(ctx context.Context, cmd *exec.Cmd) error {
  done := make(chan struct{})
 defer(close(done))
  select {
  case <- ctx.Done():
    // Send interrupt signal if not windows.
  case <- time.After(termination*time.Second):
   // Send kill signal
  case <-done
 }
 err := cmd.Wait()
 return err
}


 }
```

I think this is it. And then we can merge this in!

Thank you for your patience.

Comment (cedrickring) [reply pkg/skaffold/build/custom/custom.go]: According to your suggestion, the process will only run for `termination` seconds and then be killed. i think the nested `select` is still necessary to wait for the context to be cancelled...

Comment (tejal29) [reply pkg/skaffold/build/custom/custom.go]: Sounds good.  Running kokoro tests now

**Pull request merged**